File: 03/d03-1.f90
program d03_1
  implicit none

  integer :: fd, iostat
  character(100) :: buffer
  logical :: found1(52), found2(52)
  integer, allocatable :: priorities(:)
  integer :: total_priority

  open(newunit=fd, file="input.dat", action="read")
  total_priority = 0
  do
    read(fd, *, iostat=iostat) buffer
    if (iostat /= 0) exit
    priorities = iachar(transfer(trim(buffer), ['a']))
    where (priorities >= iachar("a"))
      priorities = priorities - iachar("a") + 1
    elsewhere
      priorities = priorities - iachar("A") + 27
    end where
    found1(:) = .false.
    found1(priorities(: size(priorities) / 2)) = .true.
    found2(:) = .false.
    found2(priorities(size(priorities) / 2 + 1 :)) = .true.
    print *, "PRIO:", findloc(found1 .and. found2, .true., dim=1)
    total_priority = total_priority + findloc(found1 .and. found2, .true., dim=1)
  end do
  close(fd)
  print "(a, i0)", "Total priority of double items: ", total_priority

end program d03_1

File: 03/d03-1.py
#!/usr/bin/env python
import numpy as np

total_priority = 0
fp = open("input.dat", "r")
for line in fp:
    charcodes = np.array([ord(c) for c in line.rstrip()])
    priorities = np.where(
        charcodes >= ord("a"), charcodes - ord("a"), charcodes - ord("A") + 26)
    found1 = np.zeros(52, dtype=bool)
    found1[priorities[: len(priorities) // 2]] = True
    found2 = np.zeros(52, dtype=bool)
    found2[priorities[len(priorities) // 2 :]] = True
    total_priority = total_priority + np.argwhere(np.logical_and(found1, found2))[0, 0] + 1

print("Total priority of double items: ", total_priority)
